[docs] [develop] Write-heavy workloads clean up (#15816)
yugabyte/yugabyte-db: docs/content/preview/develop/quality-of-service/write-heavy-workloads.md

### Limited disk IOPS or bandwidth

In many cloud environments, the disk IOPS and/or the network bandwidth is rate-throttled. Holistically, if there are such disk constraints (limits on IOPS / bandwidth), that results in a back-pressure across all disk writes in the system, and would therefore manifest as one of the above conditions.

## Write throttling triggers

YugabyteDB defines the following triggers for throttling incoming writes:

### Stop writes trigger

Stop writes trigger is activated in one of the following scenarios:

* **Too many SST files:** The number of SST files exceeds the value determined by the flag `sst_files_hard_limit`, which defaults to 48. After the hard limit is hit, no more writes are processed, all incoming writes are rejected.

> * **Memstores flushed too frequently:** This condition occurs if there are a large number of tables (or more accurately, a large number of tablets) all of which get writes. In such cases, the memstores are forced to flush frequently, resulting in too many SST files. In such cases, you can tune the total memstore size allocated. Total memstore size is the minimum of the two flags: `global_memstore_size_mb_max` (default value is 2GB) and `global_memstore_size_percentage` (defaults to 10% of total YB-TServer memory allocated. There are 2 different options for controlling how much memory is allocated to YB-TServer:

Reviewer: if there is a large number

> * **Memstores flushed too frequently:** This condition occurs if there are a large number of tables (or more accurately, a large number of tablets) all of which get writes. In such cases, the memstores are forced to flush frequently, resulting in too many SST files. In such cases, you can tune the total memstore size allocated. Total memstore size is the minimum of the two flags: `global_memstore_size_mb_max` (default value is 2GB) and `global_memstore_size_percentage` (defaults to 10% of total YB-TServer memory allocated. There are 2 different options for controlling how much memory is allocated to YB-TServer:

Reviewer: This condition occurs if there are a large number of tables (or more accurately, a large number of tablets)  --> I'm not sure why the same thing needs to be repeated in parentheses

Reviewer: Suggestion:

This condition occurs if there is a large number of tables that get writes. In such cases, the memstores are forced to flush frequently, resulting in too many SST files. You can tune the total memstore size allocated. Total memstore size is the minimum of the two flags: `global_memstore_size_mb_max` (default value is 2GB) and `global_memstore_size_percentage` (defaults to 10% of total YB-TServer memory allocated). You can use one of the following options to control the amount of memory allocated to YB-TServer:

>   * By setting `default_memory_limit_to_ram_ratio` to control what percentage of total RAM on the instance the process should use

Reviewer: * Set `default_memory_limit_to_ram_ratio` to control percentage of total RAM on the instance the process should use.
  * Specify an absolute value using `memory_limit_hard_bytes`. For example, to provide YB-TServer with 32GB of RAM, use `--memory_limit_hard_bytes 34359738368`.

### WAL writes are too slow

> If WAL writes are slow, then writes experience higher latency which creates a natural form of admission control. The frequency of data getting synchronized to disk is controlled by the `durable_wal_write flag`. Note that `fsync` is disabled by default, as YugabyteDB is expected to run in a Replication Factor of 3 mode where each tablet is replicated onto three independent fault domains such as hosts, availability zones, regions, or clouds. If enabled, every single write into DocDB, which is a YugabyteDB's underlying document store, will be synchronized to disk before its execution is deemed successful. This means that even though there will be an increase in safety, performance will be negatively affected.

Reviewer 2: ```suggestion
If WAL writes are slow, then writes experience higher latency which creates a natural form of admission control. The frequency of data getting synchronized to disk is controlled by the `durable_wal_write flag`. Note that `fsync` is disabled by default, as YugabyteDB is expected to run in a Replication Factor 3 mode where each tablet is replicated onto three independent fault domains such as hosts, availability zones, regions, or clouds. Enabling `fsync` means that every write into DocDB (YugabyteDB's underlying document store) must be synchronized to disk before its execution is deemed successful. This increase in safety brings a corresponding performance decrease.
```

### Limited disk IOPS or bandwidth

> In many cloud environments, the disk IOPS and network bandwidth are rate-throttled. Holistically, if there are such disk constraints (limits on IOPS and bandwidth), it results in a back-pressure across all disk writes in the system and would manifest as one of the preceding conditions.

Reviewer 2: ```suggestion
In many cloud environments, disk IOPS and network bandwidth are rate-limited. Such disk constraints (limits on IOPS and bandwidth) result in a back-pressure across all disk writes in the system, and manifest as one of the preceding conditions.
```
